sed/binning.py

```diff
@@ -355,7 +355,85 @@ def _hist_from_bin_ranges(
             j = (sample[t, i] - ranges[i, 0]) * delta[i]
             is_inside = is_inside and (0 <= j < bins[i])
             flatidx += int(j) * strides[i]
+            if not is_inside:
+                break  # don't check all axes if you already know you're out of the range
+        if is_inside:
+            Hflat[flatidx] += int(is_inside)
+
+    return H
+
+
+@numba.jit(nogil=True, parallel=False, nopython=True)
+def binsearch(bins: np.ndarray, val: float) -> int:
```

Comment:
is this similar to the already implemented `bisect_left` from the standard library (https://docs.python.org/3/library/bisect.html)?

Reply:
yep, looks exactly the same! however:

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Untyped global name 'bisect_left': Cannot determine Numba type of <class 'builtin_function_or_method'>

File "..\sed\binning.py", line 495:
def _hist_from_bins(
    <source elided>
            # j = binsearch(bins[i], sample[t, i])
            j = bisect_left(bins[i], sample[t, i])
            ^

Numba doesn't know bisect_left, so I had to do it manually.

Reply:
yeah, numba is a bit picky with that. You can try this one from numpy:
https://numpy.org/doc/stable/reference/generated/numpy.searchsorted.html

According to numba's docs, it's supported:
https://numba.readthedocs.io/en/stable/reference/numpysupported.html?highlight=numpy%20

Sorry for being harsh on this, but my point on this is to reuse well-tested functions instead of creating them by ourselves :)